sharkfuser/include/fusilli/support/asm_emitter.h

```diff
@@ -1090,27 +1052,98 @@ inline std::string ConvDGradNode::emitNodePreAsm() const {
 //
 //===----------------------------------------------------------------------===//
 
+// Get permute operations for input tensor in MLIR assembly format.
+// inputIndex: 0 for IN_0, 1 for IN_1, 2 for IN_2
+inline std::string PointwiseNode::getPermuteInputOpsAsm(int inputIndex) const {
+  // Get the input tensor based on index
+  std::shared_ptr<TensorAttr> inputT = pointwiseAttr.inputs.at(
+      static_cast<PointwiseAttr::InputNames>(inputIndex));
+  std::ostringstream oss;
+  std::string prefix = "permute_IN_" + std::to_string(inputIndex);
+  std::string suffix = pointwiseAttr.getName();
+
+  // Emit permute dimensions based on layout.
+  oss << getListOfIntOpsAsm(
+      getLogicalToPhysicalPermuteOrder(inputT->getDim(), inputT->getStride()),
```

Comment:
I'm a little confused by the naming of this. Aren't we going from physical -> logical on the inputs? 

Consider operand `bT` (IN_1) of logical dim `{128, 256}` stride `{1, 128}`.
```c++
  auto bT =
      graph->tensor(TensorAttr()
                        .setName("arg1_add_transposed")
                        .setDim({128, 256})
                        .setStride({1, 128})); // Transposed: stride={1, 128}
```

In the emitted MLIR, transpose ops are inserted to permute operand `bT` from physical dims `{256, 128}` to logical dims `{128, 256}` before applying the pointwise op, right?

Reply:
This should be called `getPhysicalToLogicalPermuteOrder` instead ([comment](https://github.com/nod-ai/shark-ai/pull/2653#discussion_r2512981432)). It makes sense because we want the ops to operate on logical dims but the main inputs/outputs are physical, so the translation from inputs -> op is physical->logical and from op -> outputs is logical -> physical (inverse).